### ISSUE TITLE
Kodi Matrix Update Issues

### ISSUE THREAD

**Issue by PtahX** [aznhusband/aznhusband.github.io#53]
I just took the matrix beta update on Kodi, many apps are no longer functioning including ICDrama, I'd suggest holding off on the update for now.

**Comment by boleanly** [comment]
Matrix uses Python 3 only and needs Python 3 compatible code. I'm not sure if Icdrama or all its prerequisites are Python 3 compatible

**Comment by PtahX** [comment]
> Matrix uses Python 3 only and needs Python 3 compatible code. I'm not sure if Icdrama or all its prerequisites are Python 3 compatible
I've no idea if anyone plans to try and tackle this one or not, but I spent some time converting the files using python 2to3 directly and uploaded them. I plan to play around but I doubt I'll be able to get it working with my very crap programming skill set lol  [(https://github.com/PtahX/icrama3conversions)]

**Comment by boleanly** [comment]
I actually got a fork of this plugin to work in the Kodi 18 Python 3 dev build using the 2to3 tool. But that fork is a bit modified compared to vanilla so I'm planning to just do the conversion again from the current commit in master.

From my testing, the 2to3 converts it pretty well. The only errors that need to be fixed are from strings being unicode by default and won't operate with byte strings without decoding/encoding them, and from Kodi 19 replacing the current method/function that the ListItem Icon and Thumbnail images are applied with just ListItem.setArt(). And then it's just cleaning up Python 3 import statements such as urllib turning into three things or 'import pickle as pickle'. Not really sure if anything new will pop up with the stable Kodi 19 build in the future though

**Comment by boleanly** [comment]
Ok, I got it working on Kodi 19 Beta 2 if you want to test it. At my fork and this branch (https://github.com/boleanly/aznhusband.github.io/tree/kodi19)

If it's all good then I can also do a pull request

**Comment by PtahX** [comment]
Awesome & thank you!  I'll give it a test run today and check it out. Very cool - that was no small task even with 2 to 3 I have been playing around with it as well but couldn't get it working. 

> Ok, I got it working on Kodi 19 Beta 2 if you want to test it. At my fork and this branch (https://github.com/boleanly/aznhusband.github.io/tree/kodi19)
> 
> If it's all good then I can also do a pull request

**Comment by PtahX** [comment]
so far the only issue I am seeing is a dependency error on the script.module.resuleurl version 5.1.0 with jsergio123 repo installed from ICDrama

` <import addon="script.module.resolveurl" version="5.0.38"/>` is still showing in the addons.xml 


I think ICDrama is still pointing to jsergios directory -  gujaj has taken over last I heard. It might be easier to include the gujal dir with it to always have the latest updates but I'm not sure if that is the best approach.

here is the code for the dir gujal repo to always have the latest updates (from [https://github.com/jsergio123/script.module.resolveur]

`<dir>
    <info compressed="false">https://raw.githubusercontent.com/Gujal00/smrzips/master/addons.xml</info>
    <checksum>https://raw.githubusercontent.com/Gujal00/smrzips/master/addons.xml.md5</checksum>
    <datadir zip="true">https://raw.githubusercontent.com/Gujal00/smrzips/master/zips/</datadir>
</dir>
`

**Comment by boleanly** [comment]
Oh yeah, I forgot to mention I didn't mess with the repo. I just zipped the plugin up and sideloaded it.

New issue though is that a Python 2 plugin doesn't show up in Kodi 19 and vice versa. That means if we want to support both Kodi 18- and Kodi 19+, we'll need to have two plugins in the repo with the minimum difference being the python dependency entry in the addon.xml requiring 2+ or 3+.

And should we make the code Python 2/3 compatible?

Edit: Woops. To clarify, I meant supporting both Kodi 18 and Kodi 19 requires two separate plugins in the repo. The only minimum difference between them is changing the Python dependency version from 2 to 3 or vice versa.

**Comment by fehmi1907** [comment]
@boleanly thanks for your work - it's great to know that the add-on is finally working on Matrix!

Scrubbed below. I had accidentally downloaded the master.

~~I have downloaded the folder https://github.com/boleanly/aznhusband.github.io/tree/kodi19/src/plugin.video.icdrama and when I try to install the ZIP, Matrix nags about Python compatibility. I believe this is related to addon.xml which says version="2.1.0" as you've mentioned above.~~

~~What would be the best way to switch to your fork of the add-on? Should I modify addon.xml? or just replacing the existing ICDrama folder in Kodi would suffice?~~

**Comment by boleanly** [comment]
@fehmi1907 The addon.xml in the plugin src/ folder should be correct.

The fork doesn't have a github page so the repo & add-on are not hosted. Installing from the repo seems to be the only way to also install updated versions of dependencies automatically. So if you're sideloading, you'll probably need to sideload the latest ResolveUrl too here: https://github.com/Gujal00/smrzips/tree/master/zips/script.module.resolveurl

**Comment by fehmi1907** [comment]
> @fehmi1907 The addon.xml in the plugin src/ folder should be correct.

~~@boleanly could you please double-check that? I see the following in addon.xml~~

`    <import addon="xbmc.python" version="2.1.0"/>`

> The fork doesn't have a github page so the repo & add-on are not hosted. Installing from the repo seems to be the only way to also install updated versions of dependencies automatically. So if you're sideloading, you'll probably need to sideload the latest ResolveUrl too here: https://github.com/Gujal00/smrzips/tree/master/zips/script.module.resolveurl

Will do - thanks for the heads up!

**Comment by boleanly** [comment]
@fehmi1907 In this file? https://github.com/boleanly/aznhusband.github.io/blob/kodi19/src/plugin.video.icdrama/addon.xml

**Comment by fehmi1907** [comment]
@boleanly I had accidentally downloaded a ZIP of the master - apologies for that! I have downloaded the Kodi19 branch and will try it out shortly.

**Comment by o6uoq** [comment]
@fehmi1907 did this work for you?

**Comment by fehmi1907** [comment]
@o6uoq yes, my wife has been happily using this on Matrix for months :)

**Comment by ameizeing** [comment]
@fehmi1907
I'm trying to get icdrama plug in which works on Kodi 19. I'm new to github. Can you pls point the way to which of @boleanly zip file to download? I downloaded "plugin.video.icdrama-boleanly.zip" but get xbmc.python 2.1.0 dependency not met error

**Comment by boleanly** [comment]
Hi @ameizeing , you should check out soraxas's versions in #55

**Comment by ameizeing** [comment]
@boleanly Thanks much for the lead! Have installed soraxa's latest release & so happy that it works on Kodi 19 😊
